### PR TITLE
Type layouts for structured buffers with counters

### PR DESCRIPTION
**Pull request by expipiplus1**

### CODE REVIEW

File: source/slang/slang-reflection-api.cpp [hunk 1]
```diff
@@ -2298,6 +2363,20 @@ SLANG_API SlangInt spReflectionTypeLayout_getFieldBindingRangeOffset(SlangReflec
...
 
+SLANG_API SlangInt spReflectionTypeLayout_getExplicitCounterBindingRangeOffset(SlangReflectionTypeLayout* inTypeLayout)
+{
+    auto typeLayout = convert(inTypeLayout);
+    if(!typeLayout) return 0;
+
+    if(const auto structuredBufferTypeLayout = as<StructuredBufferTypeLayout>(typeLayout))
+    {
+        getExtendedTypeLayout(structuredBufferTypeLayout);
+        return structuredBufferTypeLayout->counterVarLayout->bindingRangeOffset;
```
Comment (csyonghe) [inline source/slang/slang-reflection-api.cpp:2374]: What if `counterVarLayout` is nullptr? Is `counterVarLayout->bindingRangeOffset` always 1 if it exists?

File: tests/hlsl/append-structured-buffer.slang [hunk 2]
```diff
@@ -1,43 +1,64 @@
...
-// GLSL: uvec2 StructuredBuffer_GetDimensions_0()
-// GLSL: {
-// GLSL:     return uvec2(uint(appendBuffer_counter_0._data[0]), 8U);
-// GLSL: }
+// GLSL:      uvec2 StructuredBuffer_GetDimensions_0()
+// GLSL-NEXT: {
+// GLSL-NEXT:     return uvec2(uint(appendBuffer_counter_0._data[0]), 8U);
+// GLSL-NEXT: }
 
 // SPIRV: OpEntryPoint
```
Comment (csyonghe) [inline tests/hlsl/append-structured-buffer.slang:30]: Can delete this and the DISABLED_TEST lines.

**Review by csyonghe — DISMISSED** [review]

**Comment by csyonghe** [comment]
Looks like `AtomicCounterIncrement` and `AtomicCounterDecrement` insts are not handled in spirv-emit?

**Comment by expipiplus1** [comment]
Unless you can think of a neater way to approach this, then it's ready to merge. [`bfcff30` (#3269)](https://github.com/shader-slang/slang/pull/3269/commits/bfcff30aef43bbaa0491dc7433f1a9bd27d4136d)

**Review by csyonghe — APPROVED** [review]

**Pull request merged**